object-mapper.rst

```diff
@@ -0,0 +1,724 @@
+The returned value *must* be an object of the target type.
+
+::
+
+    // src/Dto/LegacyUserData.php
+    namespace App\Dto;
+
+    use App\Entity\User;
+    use Symfony\Component\ObjectMapper\Attribute\Map;
+
+    // Use a static factory method on the target User class for instantiation
+    #[Map(target: User::class, transform: [User::class, 'createFromLegacy'])]
+    class LegacyUserData
+    {
+        public int $userId = 0;
+        public string $name = '';
+    }
+
+::
+
+    // src/Entity/User.php
+    namespace App\Entity;
+    class User {
+        public string $name = '';
+        private int $legacyId = 0;
+
+        private function __construct() {} // Private constructor
+
+        public static function createFromLegacy(mixed $value, object $source): self
+        {
+            // $value is the initially created (empty) User object
+            // $source is the LegacyUserData object
+            $user = new self();
+            $user->legacyId = $source->userId;
+            // Property mapping will happen *after* this method returns $user
+            return $user;
+        }
+    }
+
+Mapping Multiple Targets
+------------------------
+
+A source class can be configured to map to multiple different target classes.
+Apply the ``#[Map]`` attribute multiple times at the class level, typically
+using the ``if`` condition to determine which target is appropriate based on the
+source object's state or other logic.
+
+::
+
+    // src/Dto/EventInput.php
+    namespace App\Dto;
+
+    use App\Entity\OnlineEvent;
+    use App\Entity\PhysicalEvent;
+    use Symfony\Component\ObjectMapper\Attribute\Map;
+
+    #[Map(target: OnlineEvent::class, if: [self::class, 'isOnline'])]
+    #[Map(target: PhysicalEvent::class, if: [self::class, 'isPhysical'])]
+    class EventInput
+    {
+        public string $type = 'online'; // e.g., 'online' or 'physical'
+        public string $title = '';
+
+        /**
+         * In class-level conditions, $value is null.
+         */
+        public static function isOnline(?mixed $value, object $source): bool
+        {
+            return $source->type === 'online';
+        }
+
+        public static function isPhysical(?mixed $value, object $source): bool
+        {
+            return $source->type === 'physical';
+        }
+    }
+
+    // src/Entity/OnlineEvent.php / PhysicalEvent.php ...
+
+    // Usage:
+    $eventInput = new EventInput();
+    $eventInput->type = 'physical';
+    $mapper = new ObjectMapper();
+    $event = $mapper->map($eventInput); // Automatically maps to PhysicalEvent
+
+Mapping Based on Target Properties (Source Mapping)
+---------------------------------------------------
+
+Sometimes, it's more convenient to define the mapping on the *target* class,
+specifying where its properties should come from in the *source* class. This is
+done using the ``source`` parameter within the ``#[Map]`` attribute on the target's
+properties.
```

Comment:
What if both the source and the target classes define #[Map] attributes? Which one takes precedence?